### PR TITLE
Support Tagging API for parameter group

### PR DESCRIPTION
**Pull request by kyriechen96**
Support Tagging API for parameter group, so customers could update tags for their parameter groups after creation.

Issue #, if available:

Description of changes: Modifies existing hooks and templates. Adds new hooks and templates to support tagging API.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/parameter_group/read_post_set_output.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	ctx context.Context,
+	resourceARN string,
+) ([]*svcapitypes.Tag, error) {
+	resp, err := rm.sdkapi.ListTagsWithContext(
+		ctx,
+		&svcsdk.ListTagsInput{
+			ResourceArn: &resourceARN,
+		},
+	)
+	rm.metrics.RecordAPICall("GET", "ListTagsForResource", err)
```
Comment (nmvk) [inline pkg/resource/parameter_group/read_post_set_output.go:19]: API name is `ListTags`

File: pkg/resource/parameter_group/update_post_set_update.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	if len(toDelete) > 0 {
+		rlog.Debug("removing tags from parameter group", "tags", toDelete)
+		_, err = rm.sdkapi.UntagResourceWithContext(
+			ctx,
+			&svcsdk.UntagResourceInput{
+				ResourceArn: arn,
+				TagKeys:     toDelete,
+			},
+		)
+		rm.metrics.RecordAPICall("UPDATE", "RemoveTagsFromResource", err)
```
Comment (nmvk) [inline pkg/resource/parameter_group/update_post_set_update.go:35]: API Name is `UnTagResource`

File: templates/hooks/parametergroup/sdk_update_pre_build_request.go.tpl [hunk 3]
```diff
@@ -4,4 +4,6 @@ if delta.DifferentAt("Spec.ParameterNameValues") {
     if ko != nil || err != nil {
         return ko, err
     }
-}
\ No newline at end of file
+}
+
+
```
Comment (nmvk) [inline templates/hooks/parametergroup/sdk_update_pre_build_request.go.tpl:9]: nit: Remove the extra lines

File: templates/hooks/parametergroup/sdk_update_post_set_output.go.tpl [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+if err != nil {
+    return nil, err
+}
+
+if delta.DifferentAt("Spec.Tags") {
+    err = rm.updateTags(ctx, ko, desired, latest);
+}
+
+if err != nil {
+    return nil, err
```
Comment (nmvk) [inline templates/hooks/parametergroup/sdk_update_post_set_output.go.tpl:12]: Why we cannot do this in `sdk_update_pre_build`?

File: pkg/resource/parameter_group/update_post_set_update.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	if len(toAdd) > 0 {
+		rlog.Debug("adding tags to parameter group", "tags", toAdd)
+		_, err = rm.sdkapi.TagResourceWithContext(
+			ctx,
+			&svcsdk.TagResourceInput{
+				ResourceArn: arn,
+				Tags:        sdkTagsFromResourceTags(toAdd),
+			},
+		)
+		rm.metrics.RecordAPICall("UPDATE", "AddTagsToResource", err)
```
Comment (nmvk) [inline pkg/resource/parameter_group/update_post_set_update.go:50]: Update API name here too

**Review by nmvk — CHANGES_REQUESTED** [review]

File: pkg/resource/parameter_group/update_post_set_update.go (hunk 5 above)
Comment (kyriechen96) [reply pkg/resource/parameter_group/update_post_set_update.go]: Done.

File: templates/hooks/parametergroup/sdk_update_post_set_output.go.tpl (hunk 4 above)
Comment (kyriechen96) [reply templates/hooks/parametergroup/sdk_update_post_set_output.go.tpl]: Yes, I just changed it.

File: templates/hooks/parametergroup/sdk_update_pre_build_request.go.tpl (hunk 3 above)
Comment (kyriechen96) [reply templates/hooks/parametergroup/sdk_update_pre_build_request.go.tpl]: Got it.

File: pkg/resource/parameter_group/update_post_set_update.go (hunk 2 above)
Comment (kyriechen96) [reply pkg/resource/parameter_group/update_post_set_update.go]: Done.

File: pkg/resource/parameter_group/read_post_set_output.go (hunk 1 above)
Comment (kyriechen96) [reply pkg/resource/parameter_group/read_post_set_output.go]: Done.

File: apis/v1alpha1/ack-generate-metadata.yaml [hunk 6]
```diff
@@ -1,13 +1,13 @@
 ack_generate_info:
-  build_date: "2022-04-15T17:21:57Z"
-  build_hash: 50c64871bcaf88b9ee200eb8d6b8245fa8f675eb
-  go_version: go1.17.5
-  version: v0.18.4
-api_directory_checksum: 6972849b7a809751f78bbafa31eeb725ed5bc126
+  build_date: "2022-10-26T20:52:18Z"
+  build_hash: 5ee0ac052c54f008dff50f6f5ebb73f2cf3a0bd7
+  go_version: go1.19
+  version: 5ee0ac0
```
Comment (A-Hilaly) [inline apis/v1alpha1/ack-generate-metadata.yaml:5]: We need to keep the `ack-generate` and `runtime` versions equal (make sure to fetch the code-generator tags `git fetch --tags --all` before the generation)

Comment (A-Hilaly) [reply apis/v1alpha1/ack-generate-metadata.yaml:5]: Ideally we should generate the controller using version `v0.20.1` after bumping the runtime version to `v0.20.1` (PR #30)

File: pkg/resource/parameter_group/post_set_output.go [hunk 7]
```diff
@@ -15,7 +15,6 @@ package parameter_group
 
 import (
 	"context"
-
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/post_set_output.go:18]: Please keep the imports organized using the Go best practices mentioned in https://github.com/golang/go/wiki/CodeReviewComments#imports

File: pkg/resource/parameter_group/read_post_set_output.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package parameter_group
+
+import (
+	"context"
+	svcapitypes "github.com/aws-controllers-k8s/memorydb-controller/apis/v1alpha1"
+	svcsdk "github.com/aws/aws-sdk-go/service/memorydb"
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/read_post_set_output.go:6]: Same comment regarding the order and the separation of the imports

File: pkg/resource/parameter_group/sdk.go [hunk 9]
```diff
@@ -18,13 +18,15 @@ package parameter_group
...
 	"errors"
+	"fmt"
 	"reflect"
 	"strings"
 
 	ackv1alpha1 "github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1"
 	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
 	ackcondition "github.com/aws-controllers-k8s/runtime/pkg/condition"
 	ackerr "github.com/aws-controllers-k8s/runtime/pkg/errors"
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/sdk.go:29]: This is a good example of how to seperate/order import paths

File: pkg/resource/parameter_group/sdk.go [hunk 10]
```diff
@@ -112,6 +118,15 @@ func (rm *resourceManager) sdkFind(
...
 	rm.setStatusDefaults(ko)
+	if ko.Status.ACKResourceMetadata != nil && ko.Status.ACKResourceMetadata.ARN != nil {
+		resourceARN := (*string)(ko.Status.ACKResourceMetadata.ARN)
+		tags, err := rm.getTags(ctx, *resourceARN)
+		if err != nil {
+			return nil, err
+		}
+		ko.Spec.Tags = tags
+	}
+
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/sdk.go:129]: There is no need to perform a nil check operation on `ko.Status.ACKResourceMetadata` and `ko.Status.ACKResourceMetadata.ARN` here. If the execution makes it to this block it means that a Parameter group exists and you'll have an ARN 100%. Check L116-L118

File: pkg/resource/parameter_group/sdk.go [hunk 11]
```diff
@@ -247,14 +264,25 @@ func (rm *resourceManager) sdkUpdate(
...
 		}
 	}
+
+	if delta.DifferentAt("Spec.Tags") {
+		err = rm.updateTags(ctx, desired, latest)
+	}
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/sdk.go:284]: The error check here should be right after L279. Otherwise, if there is no difference at `Spec.Tag` it's gonna evaluate the error declared in L264

File: pkg/resource/parameter_group/update_pre_build_request.go [hunk 12]
```diff
@@ -15,6 +15,8 @@ package parameter_group
 
 import (
 	"context"
+	svcapitypes "github.com/aws-controllers-k8s/memorydb-controller/apis/v1alpha1"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/update_pre_build_request.go:19]: Same comment on import paths

File: pkg/resource/parameter_group/update_pre_build_request.go [hunk 13]
```diff
@@ -15,6 +15,8 @@ package parameter_group
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/update_pre_build_request.go:15]: Why was this file renamed to `update_pre_build_request.go`? maybe better to just name it `hooks.go`?

File: pkg/resource/parameter_group/update_pre_build_request.go [hunk 14]
```diff
@@ -77,3 +79,88 @@ func (rm *resourceManager) resetAllParameters(
...
+func computeTagsDelta(
+	desired []*svcapitypes.Tag,
+	latest []*svcapitypes.Tag,
+) (added []*svcapitypes.Tag, removed []*string) {
+	toDelete := []*string{}
+	toAdd := []*svcapitypes.Tag{}
+
+	desiredTags := map[string]string{}
+	for _, tag := range desired {
+		desiredTags[*tag.Key] = *tag.Value
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/update_pre_build_request.go:140]: In case the user provides a desired tag key but no value, the code is going to panic here

File: templates/hooks/parametergroup/sdk_update_pre_build_request.go.tpl [hunk 15]
```diff
@@ -4,4 +4,12 @@ if delta.DifferentAt("Spec.ParameterNameValues") {
...
\ No newline at end of file
+}
+
+if delta.DifferentAt("Spec.Tags") {
+    err = rm.updateTags(ctx, desired, latest);
+}
+
+if err != nil {
+    return nil, err
+}
```
Comment (A-Hilaly) [inline templates/hooks/parametergroup/sdk_update_pre_build_request.go.tpl:15]:
```suggestion
if delta.DifferentAt("Spec.Tags") {
    err = rm.updateTags(ctx, desired, latest);
    if err != nil {
        return nil, err
    }
}
```

File: pkg/resource/parameter_group/update_pre_build_request.go [hunk 16]
```diff
@@ -77,3 +79,88 @@ func (rm *resourceManager) resetAllParameters(
 	}
 	return &resource{ko}, nil
 }
+
+func (rm *resourceManager) updateTags(
```
Comment (A-Hilaly) [inline pkg/resource/parameter_group/update_pre_build_request.go:83]: it would nice having a comment describing what this function is doing.

**Review by A-Hilaly — CHANGES_REQUESTED** [review]
Nice, thank you @kyriechen96!
I left few comments and suggestions bellow.

File: pkg/resource/parameter_group/sdk.go (hunk 9 above)
Comment (kyriechen96) [reply pkg/resource/parameter_group/sdk.go:29]: Thank you for the example! Just a quick question. For a group of imports, shouldn't them be ordered alphabetically? Like putting ackcompare before ackv1alpha1?

Comment (A-Hilaly) [reply pkg/resource/parameter_group/sdk.go:29]: Good question! It's organized alphabetically using the import path not the import alias :)
For example in `ackcondition "github.com/aws-controllers-k8s/runtime/pkg/condition"`, `ackcondition` is the import alias and `"github.com/aws-controllers-k8s/runtime/pkg/condition"` is the import path.

Dev tip: You can use [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) with your favorite editor, so that it always formats the imports for you.

Comment (kyriechen96) [reply pkg/resource/parameter_group/sdk.go:29]: Thank you for the explanation! However, I did check the doc and still feel a bit confused, because it only says adding a blank line between each group. For goimports, I tried to run "go install golang.org/x/tools/cmd/goimports@latest" in the terminal of GoLand but nothing happened.

Comment (kyriechen96) [reply pkg/resource/parameter_group/sdk.go:29]: Should I just click "sort imports"? I saw the option of it.

File: pkg/resource/parameter_group/update_pre_build_request.go (hunk 13 above)
Comment (kyriechen96) [reply pkg/resource/parameter_group/update_pre_build_request.go:15]: Because I wanted to make sure those file names match their hook names(hook call functions from file). Do you think it is better to put all functions into a hooks.go file?

Comment (kyriechen96) [reply pkg/resource/parameter_group/update_pre_build_request.go:15]: If merging all functions into a single hooks.go file, the pro is that only one hook file will show up, and the con is that it might be hard to find a specific function which is called by a specific template(hook).

Comment (kyriechen96) [reply pkg/resource/parameter_group/update_pre_build_request.go:15]: I found aws-rds team uses only one hooks.go file to include all functions. Currently aws-memorydb controller uses separate files for functions that hooks use. Should I merge them into one hooks.go file?

Comment (jaypipes) [reply pkg/resource/parameter_group/update_pre_build_request.go:15]: @kyriechen96 yeah, that's currently the practice that most of the controllers follow...a single `pkg/$resource/hooks.go` file.

Comment (kyriechen96) [reply pkg/resource/parameter_group/update_pre_build_request.go:15]: Got it!

File: pkg/resource/parameter_group/update_pre_build_request.go (hunk 14 above)
Comment (kyriechen96) [reply pkg/resource/parameter_group/update_pre_build_request.go:140]: Is a tag pair that only has key but no value acceptable? Currently, I added an If condition statement to run L140 only if value is not empty. Do I need to add an error if input tag pair has no value?

File: apis/v1alpha1/ack-generate-metadata.yaml (hunk 6 above)
Comment (kyriechen96) [reply apis/v1alpha1/ack-generate-metadata.yaml:5]: [](url)> We need to keep the `ack-generate` and `runtime` versions equal (make sure to fetch the code-generator tags `git fetch --tags --all` before the generation)

I've tried the command and regenerated, but the version still shows 5ee0ac0 in the file.

I guess this is the reason. Code-generator shows 5ee0ac0 as version on github now.
Screen Shot 2022-11-02 at 1 23 06 PM
